.travis.yml

```diff
@@ -1,45 +1,59 @@
+  - DOCKER_DIR="$HOME/.cache/docker"
+  - secure: HhT1TdJcpqys8juVMw/DIZeK7oD4595TEKH5KlowH7MvwwFAUyQFb5W63F8dgk7elvRG+3fmga/m1JfXO+Iu7PVD912eiNDagW9aB3CEl3Z8zg+JUL8IjpMCkyKQDyJMnfOkrzdxdaqfOK+WmF+13f2qBu9Kc7wdXuzgHQrg4+0= # CI_REGISTRY_USER
+  - secure: hh7VDZnkxgl/vqHtS4IpXfIAckKpVQvoCzNW7fstr5Mcu8KNiCWIPgObBRm+m13aqpcFTMWQ6lT2kzORz2wWRbDeVhI1eGWOJswGNHPHZLO0Jaei6yfY2nY2mpxZbl+vdg00jkN64mi1ab3e++QgeLFruW0gyNefXX7E5L/mHTs= # CI_REGISTRY_PASS
 
-script:
-  - SCITRAN_PERSISTENT_DB_PORT=27017 tests/bin/run-tests-ubuntu.sh
+cache:
+  directories:
+  - $DOCKER_DIR
 
-after_success:
-  - if [ "$TRAVIS_EVENT_TYPE" == "push" -o "$TRAVIS_TAG" ]; then
-        SSH_KEY_FILE=$(mktemp -p $HOME/.ssh/);
+before_install:
+- sudo apt-get update
+- sudo apt-get -y install docker-ce realpath
+- docker login -u $CI_REGISTRY_USER -p $CI_REGISTRY_PASS
 
-        openssl aes-256-cbc -K $encrypted_55750ae1fbc7_key -iv $encrypted_55750ae1fbc7_iv -in .github_deploy_key.enc -out "$SSH_KEY_FILE" -d;
+install: true
 
-        chmod 600 "$SSH_KEY_FILE" && printf "%s\n" \
-              "Host github.com" \
-              "  IdentityFile $SSH_KEY_FILE" \
-              "  LogLevel ERROR" >> ~/.ssh/config;
+script:
+- test -f "$DOCKER_DIR/image.tar" && docker load -i "$DOCKER_DIR/image.tar" || true
+- docker build -t core:base --target base .
+- docker build -t core:dist --target dist --build-arg VCS_BRANCH="$TRAVIS_BRANCH" --build-arg VCS_COMMIT="$TRAVIS_COMMIT" .
+- docker build -t core:testing --target testing .
+- docker save -o "$DOCKER_DIR/image.tar" $(docker history -q core:base | grep -v '<missing>') $(docker history -q core:dist | grep -v '<missing>')
+- ./tests/bin/docker-tests.sh --image core:testing
 
-        git config --global user.email "travis@travis-ci.org";
-        git config --global user.name "Travis CI";
-        git config --global push.default simple;
-    fi
-  - if [ "$TRAVIS_BRANCH" == "master" -o  "$TRAVIS_EVENT_TYPE" == "pull_request" ]; then
-        bash <(curl -s https://codecov.io/bash) -cF python;
-    fi
-  - if [ "$TRAVIS_TAG" ]; then
-        ./docker/build-trigger.sh Tag "$TRAVIS_TAG" "$BUILD_TRIGGER_URL";
-        ./bin/push-docs.sh "$GIT_REMOTE" tags "$TRAVIS_TAG" "Travis Core Docs Build - ${TRAVIS_BUILD_NUMBER}";
-    fi
-  - if [ "$TRAVIS_EVENT_TYPE" == "push" -a "$TRAVIS_BRANCH" == "master" ]; then
-        ./docker/build-trigger.sh Branch "$TRAVIS_BRANCH" "$BUILD_TRIGGER_URL";
-    fi
-  - if [ "$TRAVIS_EVENT_TYPE" == "push" -a -z "$TRAVIS_TAG" ]; then
-        ./bin/push-docs.sh "$GIT_REMOTE" branches "$TRAVIS_BRANCH" "Travis Core Docs Build - ${TRAVIS_BUILD_NUMBER}";
-    fi
-  
+after_success:
+- |
+  if [ "$TRAVIS_EVENT_TYPE" == "push" -o "$TRAVIS_TAG" ]; then
+      SSH_KEY_FILE=$(mktemp -p $HOME/.ssh/);
+      openssl aes-256-cbc -K $encrypted_55750ae1fbc7_key -iv $encrypted_55750ae1fbc7_iv -in .github_deploy_key.enc -out $SSH_KEY_FILE -d;
+      chmod 600 $SSH_KEY_FILE;
+      printf "%s\n" \
+        "Host github.com" \
```

Comment:
Personally, I would be a fan of moving these long multi-line bash scripts to a separate file, and invoking them. Keeps the yaml more readable.

Reply:
I don't mind having the details in here. I like the transparency of a single file.